[Python] Adds allOf/oneOf/anyOf composed models

spacether opened the pull request:
This PR adds composed schema (allOf/oneOf/anyOf) support to the python-experimental generator. All composed schema instances will now store instances of their composed schema models under the hood, then use them when we get attributes, set attributes, or convert the model instance to a dict for serialization.
Common model methods have been moved into parent classes with a net reduction of -3,800 lines

### Test Verification
In depth tests on the Dog allOf class have been added at:
- https://github.com/spacether/openapi-generator/blob/allof_anyof_oneof_addition/samples/client/petstore/python-experimental/test/test_dog.py#L33

A test showing that allOf parameters are accessible on an instance of Dog is at:
- https://github.com/spacether/openapi-generator/blob/allof_anyof_oneof_addition/samples/client/petstore/python-experimental/tests/test_deserialization.py#L120

### Circular Reference Problem and Solution
Note: class names in dataTypes must look like pet.Pet so we can load modules with lines like
`from package.models import pet`
at the top of the module, then we can use pet.Pet as the data type later in the model definition.
This is specifically needed for models with composed schemas because those models require circular references.

For example:
Animal uses a Cat discriminator, so the animal module imports cat
The cat module uses allOf Animal in its composed schema definition so it imports animal

The only way to make the circular imports work is using the below two techniques:
1. in cat load the animal module in with a try except import error catch, where if we get an import error we will set the animal variable to the partially loaded animal module in sys
2. delay our usage of the composed schemas (animal.Animal) by putting them in the model _composed_schemas method. That way, the code is not run until the method is invoked. If we put animal.Animal into Cat's module level constants, because the animal module is not yet fully loaded, animal.Animal does not exist and the code would not run.

### Related Issues
This may fix the below issues if the end users use the python-experimental generator:
- https://github.com/OpenAPITools/openapi-generator/issues/1662
- https://github.com/OpenAPITools/openapi-generator/issues/623

### Potential Future Work
If this PR is accepted, later on I can submit a PR adding a python-experimental client example for the openapi spec demonstrating that oneOf and anyOf also work.

- [X] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [X] If contributing template-only or documentation-only changes which will change sample output, [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) before.
- [X] Run the shell script(s) under `./bin/` (or Windows batch scripts under`.\bin\windows`) to update Petstore samples related to your fix. This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit, and these must match the expectations made by your contribution. You only need to run `./bin/{LANG}-petstore.sh`, `./bin/openapi3/{LANG}-petstore.sh` if updating the code or mustache templates for a language (`{LANG}`) (e.g. php, ruby, python, etc).
- [X] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `4.3.x`, `5.0.x`. Default: `master`.
- [X] Copy the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.


@taxpon (2017/07) @frol (2017/07) @mbohlool (2017/07) @cbornet (2017/09) @kenjones-cisco (2017/11) @tomplus (2018/10) @Jyhess (2019/01) @slash-arun (2019/11)

Authors of related work:
@rienafairefr per https://github.com/OpenAPITools/openapi-generator/pull/2170

ethan92429 commented:
Very excited to see this go in! Is there anything I can do to help it along?

spacether commented:
> Very excited to see this go in! Is there anything I can do to help it along?
Just waiting on a review and approval on this PR

spacether commented:
Python team:
@taxpon (2017/07) @frol (2017/07) @mbohlool (2017/07) @cbornet (2017/09) @kenjones-cisco (2017/11) @tomplus (2018/10) @Jyhess (2019/01) @slash-arun (2019/11)
Gentle reminder that this still needs a review
- Adding composed schema models adds a significant openapi spec feature to our python client ecosystem
- If it would help I can screen share with anyone to walk them through the updates.

spacether commented:
Python team:
@taxpon (2017/07) @frol (2017/07) @mbohlool (2017/07) @cbornet (2017/09) @kenjones-cisco (2017/11) @tomplus (2018/10) @Jyhess (2019/01) @slash-arun (2019/11)
Gentle reminder that this still needs a review

spacether commented on modules/openapi-generator/src/main/resources/python/README.mustache:
```diff
@@ -52,4 +52,4 @@ import {{{packageName}}}
 Please follow the [installation procedure](#installation--usage) and then run the following:
 
-{{> common_README }}
+{{> README_common }}
```
Good point, I reverted this back to common_README. I was trying to standardize mustache file names but other people are probably depending on this file being here.

spacether commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/PythonClientExperimentalCodegen.java:
```diff
@@ -788,4 +886,55 @@ public void setParameterExampleValue(CodegenParameter p) {
+
+    public String simpleModelName(String name) {
+        // this returns a model name like Cat
+        name = sanitizeName(name); // FIXME: a parameter should not be assigned. Also declare the methods parameters as 'final'.
```
Changed to use modelName

spacether commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/PythonClientExperimentalCodegen.java:
```diff
@@ -213,20 +238,75 @@ public String toDefaultValue(Schema p) {
+        for (String modelName : modelNames) {
+            imports.add(robustImport(modelName));
         }
+        return;
```
Removed this return

spacether commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/PythonClientExperimentalCodegen.java:
```diff
@@ -213,20 +238,75 @@ public String toDefaultValue(Schema p) {
+        return objs;
+    }
+
+    public void fixModelImports(Set<String> imports) {
```
made fixModelImports private

spacether commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/PythonClientExperimentalCodegen.java:
```diff
@@ -213,20 +238,75 @@ public String toDefaultValue(Schema p) {
+    }
+
+    public void fixOperationImports(Set<String> imports) {
+        // loops through imports and converts them all from 'module.Class' to 'from  models.module import module'
```
Moved the comment further down per suggestion

spacether commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/PythonClientExperimentalCodegen.java:
```diff
@@ -213,20 +238,75 @@ public String toDefaultValue(Schema p) {
+        return name.split("\\.")[1];
+    }
+
+    public void fixOperationImports(Set<String> imports) {
```
made fixOperationImports private

spacether commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/PythonClientExperimentalCodegen.java:
```diff
@@ -213,20 +238,75 @@ public String toDefaultValue(Schema p) {
+        return modelImport;
+    }
+
+    public String getClassName(String name) {
```
changed getClassName to getPythonClassName and made it private

spacether commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/PythonClientExperimentalCodegen.java:
```diff
@@ -213,20 +238,75 @@ public String toDefaultValue(Schema p) {
+        return "from " + modelPackage() + " import "+ moduleName;
+    }
+
+    public String robustImport(String name) {
```
Made robustImport private

spacether commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/PythonClientExperimentalCodegen.java:
```diff
@@ -396,17 +466,21 @@ public CodegenParameter fromRequestBody(RequestBody body, Set<String> imports, S
         CodegenProperty modelProp = fromProperty("body", realSchema);
         if (modelProp.isPrimitiveType && (modelProp.hasValidation || modelProp.isEnum)) {
             String simpleDataType = result.dataType;
-            result.isPrimitiveType = false;
```
This is intentional.
We are deleting this because in api_doc.mustache we using isModel to determine if this is a model or not. We generate models for models of type object and primitive types like string/number etc which have validations or enums. So the mustache code is not depending upon isPrimitiveType having any specific value here, so we don't need to set it anymore.

spacether replied:
The postProcessModelProperty is handling model properties  and this code is handling response bodies. 
- In model_doc.mustache we are switching on isPrimitiveType to determine if a property is a model or not
- in api_doc we are switching on isModel to determine if a property is a model or not
@jimschubert do you want me have both of these mustache files switch on isPrimitiveType?

spacether replied:
@jimschubert I changed the templates to only link to model docs for
- model docs if complexType is not null
- api docs if baseType is not null
This way we don't have to worry about isModel/isAlias etc.

spacether commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/PythonClientExperimentalCodegen.java:
```diff
@@ -65,6 +65,9 @@ public PythonClientExperimentalCodegen() {
         modelTemplateFiles.remove("model.mustache");
         modelTemplateFiles.put("python-experimental/model.mustache", ".py");
 
+        modelTestTemplateFiles.remove("model_test.mustache", ".py");
```
That sounds like a good idea. Let's keep that as a future improvement in a separate PR.

spacether commented on samples/client/petstore/python-experimental/docs/UserApi.md:
```diff
@@ -129,25 +129,25 @@ Creates list of users with given input array
 Name | Type | Description  | Notes
 ------------- | ------------- | ------------- | -------------
- **body** | [**[User]**](User.md)| List of user object |
+ **body** | [**[user.User]**](user.User.md)| List of user object |
```
This needs to be fixed, response array model links are mangled

spacether replied:
This has been fixed and now points at (User.md)

ethan92429 commented:
@spacether I just tried this PR out and I ran into the following issue. When specifying a discriminator property that is defined _in the superclass_, during deserialization, the client is unable to find the mentioned method in the list of attributes. Would there be a simple way to search through that list of attributes down to the root type? Example of a failing schema:
``` YAML
    SuperClass:
      type: object
      properties:
        myTypeDiscriminator:
          type: string
    SubClass:
      type: object
      properties:
        myTypeDiscriminator:
          type: string
      allOf:
        - $ref: '#/components/schemas/SuperClass'
    SubSubClass:
      type: object
      allOf:
        - $ref: '#/components/schemas/SubClass'
      discriminator:
        propertyName: 'myType'
        mapping:
          ...
```

The issue is that when SubSubClass attempts to resolve the myType property, it only looks in its own attribute map, and doesn't seem to look in the properties pulled in from the super. Let me know if this isn't clear enough. Certainly not a blocker - just wanted to mention this functionality is lacking.

spacether commented:
@ethan92429 thank you for letting me know about this.
Let me think about this and consider how to get it working.
My addition is not yet considering discriminator class mapping. In the future I plan on adding it.
This diff: https://github.com/OpenAPITools/openapi-generator/pull/4454
Is working on improving ancestor handling in composed schema models.

spacether commented:
> @ethan92429 thank you for letting me know about this.
> Let me think about this and consider how to get it working.
> My addition is not yet considering discriminator class mapping. In the future I plan on adding it.
> This diff: #4454
> Is working on improving ancestor handling in composed schema models.

@ethan92429 I got it working and added a Child composed schema model showing it working at:
- tests https://github.com/spacether/openapi-generator/blob/allof_anyof_oneof_addition/samples/client/petstore/python-experimental/test/test_child.py#L36
- spec definition https://github.com/spacether/openapi-generator/blob/allof_anyof_oneof_addition/modules/openapi-generator/src/test/resources/2_0/python-client-experimental/petstore-with-fake-endpoints-models-for-testing.yaml#L2058

ethan92429 commented:
Hi @spacether glad to hear that you got this working - just to clarify the example that I couldn't get to work (I'm trying again with the example specified in the OAS)
``` YAML
components:
  schemas:
    Animal:
      type: object
      required:
      - pet_type
      properties:
        pet_type:
          type: string
    Pet:
      type: object
      allOf:
      - $ref: '#/components/schemas/Animal'
      discriminator:
        propertyName: pet_type
        mapping:
          cachorro: Dog
    Cat:
      allOf:
      - $ref: '#/components/schemas/Pet'
      - type: object
        # all other properties specific to a `Cat`
        properties:
          name:
            type: string
    Dog:
      allOf:
      - $ref: '#/components/schemas/Pet'
      - type: object
        # all other properties specific to a `Dog`
        properties:
          bark:
            type: string
    Lizard:
      allOf:
      - $ref: '#/components/schemas/Pet'
      - type: object
        # all other properties specific to a `Lizard`
        properties:
          lovesRocks:
            type: boolean
```

Basically, I have one single discriminator defined in the base class of my deeply nested polymorphic tree, and I need all the leaves to use that deep root as their discriminator, but the Pet model can't find it because it is an inherited property. Put simply, I need inherited properties to be considered as valid discriminator mappers if they are included in allOf. Does that clarify?

spacether commented:
@ethan92429 your example is not working because of a bug in swagger-parser. I have open a ticket to track that bug here: https://github.com/swagger-api/swagger-parser/issues/1269
I am still working on a way to get your example to work.

spacether commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java:
```diff
@@ -1807,17 +1807,22 @@ public CodegenModel fromModel(String name, Schema schema) {
+                        if (m.discriminator == null && innerSchema.getDiscriminator() != null) {
                             LOGGER.debug("discriminator is set to null (not correctly set earlier): {}", name);
-                            m.discriminator = createDiscriminator(name, schema);
+                            m.discriminator = createDiscriminator(name, innerSchema);
```
Per here: https://swagger.io/specification/#discriminatorObject
We should be able to set the discriminator on the parent composed schema, but that's not yet working because of this bug in swagger parser: https://github.com/swagger-api/swagger-parser/issues/1269
In place of having discriminator at the root level of a composed schema, we should be able to use it from our definition in our allOf schemas, so I updated this code to set the discriminator on self if one of its ancestor allOf schemas define discriminator.
Please see my ParentPet class which demonstrates this usage at: https://github.com/spacether/openapi-generator/blob/allof_anyof_oneof_addition/modules/openapi-generator/src/test/resources/2_0/python-client-experimental/petstore-with-fake-endpoints-models-for-testing.yaml#L2072

NOTE: the original invocation here `m.discriminator = createDiscriminator(name, schema);` did not do anything because we already ran that exact line above in line 1772.

jimschubert replied:
I think this seems reasonable, but I'd like to have someone else on the @OpenAPITools/generator-core-team review as well since it affects all generators.

spacether commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java:
```diff
@@ -1807,17 +1807,22 @@ public CodegenModel fromModel(String name, Schema schema) {
                         }
+                        if (modelDiscriminators > 1) {
+                            LOGGER.error("Allof composed schema is inherriting >1 descriminator. Only use one discriminator: {}", composed);
+                        }
```
We should error out if the user tries to define multiple discriminators in one schema.

spacether commented:
@ethan92429 in this branch I:
- got your example working
- added it to the python-experimental sample spec https://github.com/spacether/openapi-generator/blob/allof_anyof_oneof_addition/modules/openapi-generator/src/test/resources/2_0/python-client-experimental/petstore-with-fake-endpoints-models-for-testing.yaml#L2072
- added a test of it https://github.com/spacether/openapi-generator/blob/allof_anyof_oneof_addition/samples/client/petstore/python-experimental/tests/test_deserialization.py#L124

spacether commented:
@jimschubert can you give this another look? Your feedback has been addressed and all tests are passing.

ethan92429 commented:
@spacether I've been playing around with python-experimental and ran into a hitch with using it for my spec. The problem is that with the changes to have the class property types used directly, circular type referencing or recursive definitions doesn't seem to work anymore:
``` yaml
components:
  schemas:
    Dog:
      properties:
        enemy:
          type: object
          $ref: '#/components/schemas/Dog'
```
This should be a valid spec, but the generated python attempts to reference the Dog class in the definition of the Dog class - which clearly doesn't work.  This causes a circular import. This looks to be a regression from just the plain python generator - which uses strings to define openApi types. What are your thoughts on this?

spacether commented:
@ethan92429 can you file a bug on your use case where a model has a property which is of type self? 
I can fix it by storing a model's parameter types as a class method rather than a dict. Then they will not be loaded at module load time.
